### ISSUE TITLE
[data grid] Allow for expressing the grid height in number of visible rows

### ISSUE THREAD

**Issue by Janpot** [mui/mui-x#13143]
### Summary

Looking for a way to set the height of the grid based on the desired amount of rows to be visible at once. A `heightInRows` _(better name required)_ prop that takes an amount of rows and sizes the grid to accommodate for this exact amount. To avoid a small space underneath the last row.

This would play well with the `autoPageSize` property to have a grid that sizes and paginates on a predefined amount of rows. `autoHeight` together with pagination is not enough as it can't account for height during loading, or for fewer items on the last page.

Perhaps it would even make sense to overload the `autoHeight` property to take a number?

### Examples

I guess the best analogy would be the [`rows`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/textarea#rows) property of the native `textarea` 

**Search keywords**: rows height

**Comment by michelengelen** [comment]
Added it to the board! 👍🏼

**Comment by michelengelen** [comment]
This is feasible now. I verified an implementation path, and a workaround that works today:

```tsx
const N = 5;

<DataGrid
  autoPageSize
  sx={{
    height: 'auto',
    '& .MuiDataGrid-main': {
      flexGrow: 0,
      height: `calc(var(--DataGrid-topContainerHeight) + ${N} * var(--height)
        + var(--DataGrid-bottomContainerHeight)
        + var(--DataGrid-hasScrollX) * var(--DataGrid-scrollbarSize))`,
    },
  }}
/>
```

The grid shows exactly N rows. The height stays stable during loading and on short last pages. autoPageSize resolves to exactly N, with no leftover space. Headers, pinned rows, density, and the horizontal scrollbar are covered. Note that these CSS variables are internal today.

@Janpot do you still need a built-in prop? The issue has no upvotes since 2024. If demand stays low, we would document a recipe instead.

**Comment by Janpot** [comment]
Would that work with variable row height?

**Comment by michelengelen** [comment]
No, this is solely for known row heights. IMO "fixed height" and "exactly N rows of unknown height" are contradictory requirements. The "exactly N rows of unknown height" case is already covered by the `pageSize={N}` prop.

**Comment by michelengelen** [comment]
Since this ask is kind of contradictory as stated above I will close this one. Although I think the remaining bit is a duplicate of #13263
